internal/ingress/controller/kong.go

```diff
@@ -72,10 +93,24 @@ func (n *KongController) OnUpdate(ctx context.Context, state *kongstate.KongStat
 	} else {
 		err = n.onUpdateDBMode(targetContent)
 	}
+	var target []byte
+	if n.cfg.LogSensitiveConfig {
+		target, _ = json.Marshal(targetContent)
+	} else {
+		state.Sanitize()
+		sanitizedContent := n.toDeckContent(ctx, state)
+		target, _ = json.Marshal(sanitizedContent)
+	}
 	if err != nil {
+		if n.cfg.LogLevel == "debug" {
```

Comment:
I believe (in disagreement with @hbagdi's https://github.com/Kong/kubernetes-ingress-controller/pull/991#issuecomment-749663141) that this functionality (of adding a new KIC output) should not be correlated with logging in any way because:
- [12factor](https://12factor.net/logs) says the following about logging:
    > A twelve-factor app never concerns itself with routing or storage of its output stream. It should not attempt to write to or manage logfiles. Instead, each running process writes its event stream, unbuffered, to `stdout`.
    
    Any attempt at being concerned with where and how the logs are stored is a violation of 12factor.
- It's an anti-pattern for containers to log to the local filesystem
- All features (other than being likely useful with a detailed logging level) of this functionality are orthogonal to the log stream produced by the application.
- While logging is a standard, non-unique operational concern (like metrics, GC, etc.) providing a dump from the business logic is not. Therefore I'm tempted to consider dumping the diff an auxiliary feature of the application.

Therefore, I propose the following flag set to control this feature:
```go
flag.String("failure-dump-dir", "", "If set to a directory, the ingress controller will, in case of failed application, save the generated deck configuration and the diff to the specified directory.")
flag.Bool("failure-dump-sanitized", true, "If true, the ingress controller will do the best effort to redact sensitive contents from the files saved to --failure-dump-dir.")
```

This solution brings the following benefits:
- the user specifying a destination dir enables the feature (no need to hardcode the temp dir)
- it's impossible to accidentally spam the local FS by changing log level

Reply:
I don't particularly care what we call it. da2c974 makes it a string flag that lets you select whether it includes sensitive data.

Do you indeed think writing to the filesystem is indeed likely to cause issues in practice? Again, this can't create more than one set of dump files because it's tied to the container master process, and if that dies, the container filesystem dies with it. I don't see how choosing a location makes any difference unless we expect user to mount a dedicated volume for it. What we're writing here doesn't warrant that overhead.

A possibly better solution would be to provide a download endpoint on [the status server](https://github.com/Kong/kubernetes-ingress-controller/blob/v1.1.1/cli/ingress-controller/main.go#L647-L714), but we don't have any way to expose data from the sync loop to it, and I don't think this warrants building one.

Dumping to the filesystem isn't perfect, but it's simple and should be sufficient for this.

Reply:
> I don't particularly care what we call it

My point is about separating these two concerns:
- logging (which is well defined in cloud-native best practices, and the subject feature does IMO clearly not belong)
- writing intermediate results to the FS

Naming is merely a consequence of being careful about separation of concerns (vide: "ubiquituous language" in DDD).

Why I think it's important to maintain these concerns as separate:
- "Logging", like I said before, is well defined for cloud-native applications. Specifically, raising the logging level is widely expected to increase the volume of stdout, not to produce extra files. Sure, we can nurture the "tribal knowledge" that "KIC is special in that setting the log level to debug has this and that side effect", but such things add up over time as technical debt.
- Doing otherwise is at odds with several points of SOLID.
- Implementing this feature separately from logging limits points of integration between this feature and the rest of the system to the `KongState` and `file.Content` objects. Keeping the size of the integration minimal is highly desirable. Doing otherwise couples this feature with logging, which increases the size of the integration.

> Do you indeed think writing to the filesystem is indeed likely to cause issues in practice?

If we make writing to the filesystem a side effect of increasing the logging level, I think it will cause the following issues:
- violate the cloud-native contract of what logging is (vide the 12factor link above)
- introduce a non-0 probability of data leak and full disk issues
- make logging and the control loop harder to maintain

> Again, this can't create more than one set of dump files because it's tied to the container master process, and if that dies, the container filesystem dies with it.

This is not incorrect, but it depends on two further unrelated facts:
- that the Dockerfile puts KIC as PID 1 in the container,
- that Kubernetes container lifecycle deletes and recreates a container on restart (unlike docker)

Not depending on this behavior removes a whole class of problems.

> I don't see how choosing a location makes any difference unless we expect user to mount a dedicated volume for it

Making environment-specific parameters external is another principle of [12factor](https://12factor.net/config). Rather than speculating whether the user will use this in a volume or not, making this external effectively removes the question from our list of worries altogether.

> A possibly better solution would be to provide a download endpoint on the status server

I think that writing to the FS is just fine - it's simple and solves the problem. What makes a difference, though, is _how_ we implement that (from a software architecture standpoint). Which is my key point in this review.

To summarize all the points above, my distilled requests are:
- Treat generated config dumping as functionality that's minimally related to other areas of concern in KIC (such as logging).
- If there's no good reason to do otherwise, follow general SW design guidance such as SOLID, 12factor.
- Minimize interactions between different areas of concern (some examples given above)

I believe that the proposal from https://github.com/Kong/kubernetes-ingress-controller/pull/991#discussion_r564687599 is an easy way to achieve that.